### ISSUE TITLE
Support data-driven auth methods, e.g. SAML

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3670]
The server is gaining a new API feature to let `/server_settings` describe auth methods with a bit more detail, making it possible to use auth methods that aren't in a hardcoded list in the client code.

The immediate use case for this is SAML authentication, where there's a single protocol that may have any number of providers with different names, so that it makes less sense than it already does for the client to have to hardcode the possible names in advance.

The server-side implementation just landed as zulip/zulip#13307. Docs are said to be coming soon; previous discussion [in chat here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/SAML.20authentication/near/790808).

It would be good for us to implement our end soon, in order to better end-to-end validate the API choices before they're made permanent by a server release.

We'll need to keep the existing code for backwards-compatibility with older Zulip servers, but should plan to ignore the old data if the new key is present in the `/server_settings` data.

/cc @timabbott @mateuszmandera

**Comment by timabbott** [comment]
The docs updates were merged in https://github.com/zulip/zulip/pull/13339.

**Comment by gnprice** [comment]
I'm going through this now to implement the new part of the `server_settings` API -- just for the existing methods like Google and GitHub auth -- since that seems to be up now on chat.zulip.org.

@timabbott what's a way we can test actual SAML auth? I think you said something about a test machine on a DigitalOcean droplet.

Some thoughts on the API as I have my hands in it:
* The name `social_backends` for this property is pretty odd:
  * These aren't particularly "social". Google and GitHub auth are marginally so; auth with AzureAD, or `remoteuser` SSO, or a SAML provider like Okta, is pretty much exclusively about authenticating as a corporate employee, or another institutional affiliate.
    * I know there's a library we're using with a name like "django-social-auth", but that inaptly-named dependency within the server seems like very much an implementation detail from any client's perspective.
  * These aren't "backends" in any obviously meaningful sense. They're methods for authenticating, or providers of authentication or identity... but the very next thing we're going to do with any of these is go make an HTTP request from a browser. Which will typically redirect to some other site, for the user to go enter their credentials into a web page. None of that seems particularly more backend-y than basically anything else a user might do on the Internet.
    * I know Django has a concept with a name like "auth backend", and each of these might correspond to one of those. But to the extent that's a good name at all, it's a good name *for the backend* of how the given authentication method works -- i.e. the part within the server, and perhaps the server's interaction behind the scenes with the auth/identity provider's server. It doesn't make much sense in the API for clients.
  * Missing from the name: any hint that this is about authentication at all.
    * Except to the extent that "social" gives that hint? Or that someone really steeped in Django auth integrations could perhaps hazard a guess from "backends" :wink: 
    * But Zulip itself is rather more social than any of the things that can appear here. So really once you know this is coming from a Zulip server, the name carries very little information at all.

* Not sure what a good name would be. It's *almost* just a list of all authentication methods, right? It's all of them but dev, password, and LDAP... it's all except those with two particular, well-specified login flows (namely trivial, and username+password)... it's the ones that are "generic", just go to some URL... it's the ones that are opaque to the client, and all it can know is to send the user in a browser to some URL.
  * Maybe `generic_authentication`? `web_authentication_methods`?

... OK, I have some smaller things to mention, but hitting send first.

**Comment by gnprice** [comment]
Smaller thoughts on the API:
* Docs say `display_icon`, but example and czo say `display_logo`.
  * I'm inclined to say "icon" is the better term -- a "logo" is a possible choice for an icon but some icons aren't logos. What's the "logo" for `remoteuser` auth?

* Docs don't mention `name`.

* The `google` method has `"login_url": "/accounts/login/google/"`. The notes I have in a comment say about that URL:
```
    // Server versions through 2.0 accept only this URL for Google auth.
    // Since server commit 2.0.0-2478-ga43b231f9 , both this URL and the new
    // accounts/login/social/google are accepted; see zulip/zulip#13081 .
```
Should probably give the new URL, no?

* It's a little confusing having a field called `sort_order` and having it sort in reverse. Why not use `sort_order` values that have the same order they're meant to sort in?
  * Or if there's a strong desire to use bigger numbers for the first ones to show... perhaps call it something like `priority`? Or `display_priority`.

**Comment by mateuszmandera** [comment]
@gnprice @timabbott 

> Not sure what a good name would be. It's almost just a list of all authentication methods, right? It's all of them but dev, password, and LDAP... it's all except those with two particular, well-specified login flows (namely trivial, and username+password)... it's the ones that are "generic", just go to some URL... it's the ones that are opaque to the client, and all it can know is to send the user in a browser to some URL.

``RemoteUser`` is not a social backend (and it shouldn't be showing in the ``social_backends`` list at /server_settings - if it is showing for you, that'd be a weird bug :thinking: )

What these "social backends" all have in common is that the authentication happens with an external provider (rather than something on the zulip server). And all such auth methods that use an external provider should show up  in the ``social_backends`` list, even if in the future we add some that for their implementation we use a different library than ``python-social-auth``.

Thus, perhaps ``external_authentication_methods`` would be a good name? I feel like ``generic_authentication`` is too... well, "generic"  - since these auth methods do have this "externality" as their central defining characteristic :smile:

**Comment by mateuszmandera** [comment]
> Docs don't mention name

Hmm, I think this doesn't play any useful role right now, perhaps it's best to just remove this field from the API dicts.

**Comment by mateuszmandera** [comment]
>     * The `google` method has `"login_url": "/accounts/login/google/"`.
> 
> Should probably give the new URL, no?

Oh, this is a weird result of the following django urls setup:
```
    url(r'^accounts/login/social/([\w,-]+)$', zerver.views.auth.start_social_login,
        name='login-social'),
    url(r'^accounts/login/social/([\w,-]+)/([\w,-]+)$', zerver.views.auth.start_social_login,
        name='login-social-extra-arg'),
    # Backwards-compatibility (legacy) Google auth URL for the mobile
    # apps; see https://github.com/zulip/zulip/issues/13081 for
    # background.  We can remove this once older versions of the
    # mobile app are no longer present in the wild.
    url(r'^accounts/login/(google)/$', zerver.views.auth.start_social_login,
        name='login-social'),
```

combined with this method of constructing the login url for each backend for the api dicts:
``login_url=reverse('login-social', args=(social_backend.name,)),``
which translates the case with ``social_backend.name = 'google'`` to the old endpoint

Fixed in https://github.com/zulip/zulip/pull/13356

**Comment by timabbott** [comment]
> It's a little confusing having a field called sort_order and having it sort in reverse. Why not use sort_order values that have the same order they're meant to sort in?

Agreed.  @mateuszmandera I think it probably makes sense to just flip the direction for these in the code, so that lower numbers appear first.

> I'm inclined to say "icon" is the better term -- a "logo" is a possible choice for an icon but some icons aren't logos.

Agreed s/logo/icon/ would be an improvement.  @mateuszmandera can you switch this?

> The name social_backends for this property is pretty odd:

We should also keep in mind the existing legacy `authentication_methods` field, which I imagine we'll remove in a year or two but will be with us until then.  I'm not sure what a good name is either :/.

My best idea for that name is `authentication_buttons`, since that describes clearly that this is the set of third-party authentication option buttons to render, and if look at the fields of this structure, they detail precisely how to render those buttons and nothing else.  The main concern with that approach is it may be less extensible long-term if we have non-button authentication items.  

I suppose we could call them `authentication_widgets`, and include a `type: button` field on the items, for future-proofing the possibility of adding additional authentication widgets that aren't precisely a button to this data structure.  (And then the mobile code could just ignore ones that aren't `type: button`).  I think I like that option best of the ones we've proposed.

Thoughts?

**Comment by gnprice** [comment]
> Thus, perhaps `external_authentication_methods` would be a good name? I feel like `generic_authentication` is too... well, "generic" - since these auth methods do have this "externality" as their central defining characteristic

That'd be pretty reasonable!

I'd like to probe a little more the `accounts/login/sso/` case, though. From the mobile app's perspective, this actually works 100% exactly the same as all these other auth methods:
* The client doesn't need to know anything about how to do it, except a URL.
* It sends the user to that URL in a browser.
* The server does whatever it wants to do -- in most cases redirect to some other service that in turn does whatever flow it wants to do, or in this case something much simpler. But in all cases, the Zulip client knows nothing about the details.
* If the server determines auth succeeded, it sends the user back to the client, using a `zulip://` URL and the `mobile_flow_otp` one-time-pad.

So for us it'd be convenient if it did get included in the list. I wasn't observing that, but czo (the server I had handy that has this feature in the first place) doesn't have `REMOTE_USER` auth set up anyway, so that's what I expected there :slightly_smiling_face: 

Moreover I think the description I just gave of what these all have in common would make sense, except the last bullet, for a lot of clients -- any client that might not know up front the details of all possible auth flows. And the last bullet seems like it could be adapted as needed. So I think it's not only mobile, but Zulip clients in general that want to support not-just-password auth at all, that would benefit from having `REMOTE_USER` auth included in the list the same way as the others we're discussing.

**Comment by gnprice** [comment]
> having `REMOTE_USER` auth included in the list the same way as the others we're discussing.

This wouldn't necessarily make `external_authentication_methods` an inapt name, either:
* it might be the only one for a long time that doesn't involve redirecting to some other ("external") service;
* even though it doesn't redirect, this method is "external" in another sense: the client doesn't know any of the details, and has to rely on whatever's found at some URL (something "external" to the Zulip client) to drive the flow.

**Comment by gnprice** [comment]
> My best idea for that name is `authentication_buttons`, since that describes clearly that this is the set of third-party authentication option buttons to render, and if look at the fields of this structure, they detail precisely how to render those buttons and nothing else. The main concern with that approach is it may be less extensible long-term if we have non-button authentication items.
> 
> I suppose we could call them `authentication_widgets`, and include a `type: button` field on the items, for future-proofing the possibility of adding additional authentication widgets that aren't precisely a button to this data structure.

Hmm. This feels like it's awfully UI-oriented. Which is fine for something aimed at the web frontend, but for something in the API for other clients to use which have their own UIs it doesn't seem like a great fit.

For example, already today on mobile if you go to log into a realm that has both password auth and something else, we show you a button for password auth -- it's no different from the third-party auth methods in that respect. On web (at least desktop web) it works well to show the email+password fields right there at the same time as the buttons for alternate methods, but on mobile that's too crowded.

It's true that these have some details like icon and display name that are meant for presentation to the user; but those are at a similar level of abstraction to a user's name and avatar, and are similarly flexible to fit naturally in a variety of UI contexts.

**Comment by gnprice** [comment]
> > Docs don't mention name
> 
> The only role this plays right now is that if someone configures multiple SAML IdPs without setting their `display_name`s, they'll all end up with it set to `SAML` and `name` just is a key to distinguish between them. But maybe this field should just be deleted from these API dicts, since it doesn't really _do_ anything.

Yeah, not sure it's essential -- I don't expect we'd do much with it in the mobile app, either.

One thing I could see it being potentially useful for is if the client does have (or thinks it has?) any special knowledge about some of the options. For example maybe it sees there's Google auth, and it's the Android app in some future where we've found a way to successfully use the nice system flow for authenticating with the Google account the device is logged into. Or maybe something more boring, like a tweak to the presentation or just a workaround for some bug. If a client does ever for some reason want a special case like one of these, `name: 'google'` is likely a much better way to target it than testing the `login_url` or something.

**Comment by timabbott** [comment]
> Docs don't mention name

Yeah, agreed, we want the field because it provides opportunity just what Greg describes.

On `authentication_widgets`:
> Hmm. This feels like it's awfully UI-oriented. Which is fine for something aimed at the web frontend, but for something in the API for other clients to use which have their own UIs it doesn't seem like a great fit.
> This wouldn't necessarily make external_authentication_methods an inapt name, either:

So, the reason I was thinking something more like `authentication_widgets`, is that it's theoretically possible we might want to later migrate `password` authentication to that system (basically moving all the other password/LDAP related keywords into that object), so that it can e.g. benefit from the `sort_order` feature.  And `external_authentication_methods` would be a poor name for a thing that included the internal authentication options.

Really, I want to call it `authentication_methods`.  I suppose we could add an argument one must include in the `/server_settings` request to ask for a new format if we want to transition away from `external_authentication_methods` down the line.  So I'm fine with that option if folks prefer.

For the `REMOTE_USER` backend:
> So for us it'd be convenient if it did get included in the list. I wasn't observing that, but czo (the server I had handy that has this feature in the first place) doesn't have REMOTE_USER auth set up anyway, so that's what I expected there slightly_smiling_face

I think I agree it should be in this data structure; that makes live easier for mobile and terminal apps.  And actually @mateuszmandera if you look at `zerver/views/login.html`, you can see that the structure of the "Login in with SSO" button that `only_sso` provides has more or less the same structure as a SAML backend button.  Which suggests to me that setting this up the same way the SAML ones work would clean up that template significantly.  I think you should be able to test such a migration in a test production system with https://zulip.readthedocs.io/en/latest/production/authentication-methods.html#apache-based-sso-with-remote-user; you'll just want to configure Apache digest/basic auth to make it work.  

Schedule-wise I'd prioritize the other format changes we've been discussing before this, though, since this will have the least impact on the testing experience.

> sort order

Finally, I realized we should drop `sort_order` from the objects; we're sending a list of dicts, and we can just sort it before sending to the client.

**Comment by gnprice** [comment]
> Finally, I realized we should drop `sort_order` from the objects; we're sending a list of dicts, and we can just sort it before sending to the client.

Hmm indeed! Good idea. :slightly_smiling_face: 


> Really, I want to call it `authentication_methods`. I suppose we could add an argument one must include in the `/server_settings` request to ask for a new format if we want to transition away from `external_authentication_methods` down the line. So I'm fine with that option if folks prefer.

That sounds like a good plan to me.

**Comment by mateuszmandera** [comment]
> > Really, I want to call it `authentication_methods`. I suppose we could add an argument one must include in the `/server_settings` request to ask for a new format if we want to transition away from `external_authentication_methods` down the line. So I'm fine with that option if folks prefer.
> 
> That sounds like a good plan to me.

To clarify, what is the first name change we want to go with then?  ``external_authentication_methods`` instead of ``social_backends``?

And in the (possible) future move that to ``authentication_methods`` in a new format, requested through an additional argument, where it'll include all the backends?

**Comment by timabbott** [comment]
Yeah, let's go with `external_authentication_methods` for now.

There's a possible future revised version where we don't even send the new data structure unless `extended_authentication_methods=True` is passed, and if it is, we just send it under the name `authentication_methods`.  But we can't go there until we have a plan for migrating the communication about the non-external authentication methods via this data structure, which will require significant design work for uncertain gains, so I think it makes sense to not do any active work in that direction at this time.

**Comment by timabbott** [comment]
OK, we just merged #13361, which executes the main API changes discussed in this thread:
* `sort_order` is removed from the object and instead just used for sorting :)
* The objects are now called `external_authentication_methods`
* `display_logo` -> `display_icon`
* Documented the `name` field.

I think those are all the easy changes.  We should try to get converting `RemoteUserBackend` to this framework done next, but probably mobile folks can again proceed with getting their changes working under the assumption we'll succeed at that.

**Comment by gnprice** [comment]
> OK, we just merged [zulip/zulip#13361](https://github.com/zulip/zulip/pull/13361), which executes the main API changes discussed in this thread:

Awesome -- thanks to you both!

One small further point I just noticed:
* For `display_icon`, when there's no icon it'd be cleaner to indicate that with the value `null` rather than an empty string.
  * In Python this corresponds to using `None` vs. an empty string.

**Comment by timabbott** [comment]
zulip/zulip#13368 should merge with that change shortly.

**Comment by gnprice** [comment]
> We should try to get converting `RemoteUserBackend` to this framework done next, but probably mobile folks can again proceed with getting their changes working under the assumption we'll succeed at that.

@mateuszmandera Do you have time to try moving that over, so that `remoteuser` appears in `external_authentication_methods` when enabled?

If we move it after 2.1 is out, that'll be OK but will just add a few more lines of compatibility code we'll need to carry in the client for as long as 2.1 is around.

**Comment by mateuszmandera** [comment]
@gnprice I'll start working on it

**Comment by gnprice** [comment]
@mateuszmandera Excellent, thanks!

Meanwhile I've just merged (with #3675) an implementation of this feature.

Once the `remoteuser` change is in, I'll make the small mobile-side adjustment for that. The mobile code that's now in master (after #3675) will treat `remoteuser` the same way older mobile versions do, so it doesn't add any new compatibility constraints (even if we make a mobile release with this version of the code.)
